### PR TITLE
fix(docs): avoid duplicating manual documentation

### PR DESCRIPTION
**Pull request by rwaskiewicz**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features) - https://github.com/ionic-team/stencil-site/pull/928
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Unit tests (`npm test`) were run locally and passed
- [x] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

A user may have manually written text in their component's generated markdown file.
When there is no JSDoc present at the class level of a Stencil component, that manually
written content is assigned to the `docs` field of our JsDoc representation.
The original implementation would then lead to a duplication of this text, once in the 'autogenerated'
section and also above it.

GitHub Issue Number: https://github.com/ionic-team/stencil/issues/3762

## What is the new behavior?

this commit fixes a regression that was introduced in https://github.com/ionic-team/stencil/commit/2db4f4de62c9547ccafb1d382130fcf82fd9ebf4
(https://github.com/ionic-team/stencil/pull/3635). the original commit did not take into account the fact that a
user could have manually written text in their component's markdown
file. that content would be assigned to the value used to generated the
markdown file's "Overview" section, causing the content to be displayed
twice.

the contents of the `docs` field (i.e. the way that value is generated).
cannot be changed, as `JsonDocs` publicly exposes that field (via a
`JsonDocsComponent` value), and may be used by custom documentation
generators.

instead, this commit adds an optional 'overview' field to the
`JsonDocsComponent` type. the intent here is to avoid _most_
assignability errors by making this field optional. the value is
assigned when stencil generates its documentation data with the text
from the class component's jsdoc.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

This commit can be manually tested in addition to the unit tests

### Setup
1. Check out this branch, and build it with `npm run clean && npm ci && npm run build && npm pack`
2. In `/tmp/`, run `npm init stencil component test-readme && cd $_ && npm i && git init && git add -A && git commit -m 'init with cli'`
3. Install the tarball to your project: `npm i PATH_TO_TARBALL`
4. Run the build to get a baseline: `npm run build`
5. Commit your changes `git commit -am 'build with custom tarball'`

### Overview from JSDoc is Added to MD
1. Run the steps in 'Setup'
2. Add the following JSDoc to the boilerplate component `src/components/my-component/my-component.tsx`
```diff
+ /**
+ * Hello world
+ * 
+ * I am text
+ */
@Component({
```
3. Run `npm run build`
4. Note the JSDoc is properly set in `src/components/my-component/readme.md`
```diff
<!-- Auto Generated Below -->


## Overview

+ Hello world
+ 
+ I am text
+
## Properties
```
5. `git reset --hard` to restore the baseline

### Overview from MD is Not Duplicated in MD
1. Run the steps in 'Setup'
2. Add the following MD to `src/components/my-component/readme.md`
```diff
# my-component

+ This is my text
+
+ This is manually inserted
+
<!-- Auto Generated Below -->
```
3. `npm run build`
4. Note the JSDoc is NOT duplicated
5. `git reset --hard` to restore the baseline

### Overview from JsDoc is used when MD present
1. Run the steps in 'Setup'
2. Add the following MD to `src/components/my-component/readme.md`
```diff
# my-component

+ This is my text
+
+ This is manually inserted
+
<!-- Auto Generated Below -->
```
3. Add the following JSDoc to the boilerplate component `src/components/my-component/my-component.tsx`
```diff
+ /**
+ * Hello world
+ * 
+ * I am text
+ */
@Component({
```
4. Run `npm run build`
5. Note the JSDoc is properly set in `src/components/my-component/readme.md`
```diff
<!-- Auto Generated Below -->


## Overview

+ Hello world
+ 
+ I am text
+
## Properties
```
5. `git reset --hard` to restore the baseline
## Other information

There are two commits in this PR
- The first is the true bugfix
- The second is addition of more tests

### CODE REVIEW

**Review by tanner-reits — APPROVED** [review]
lgtm!

File: src/declarations/stencil-public-docs.ts [hunk 1]
```diff
@@ -19,6 +19,10 @@ export interface JsonDocsComponent {
   readme: string;
   docs: string;
   docsTags: JsonDocsTag[];
+  /**
+   * The comment found at in a class-level JSDoc for a Stencil component.
```
Comment (alicewriteswrongs) [inline src/declarations/stencil-public-docs.ts:23]: maybe

```suggestion
   * The text from the class-level JSDoc for a Stencil component, if present.
```

**Review by alicewriteswrongs — APPROVED** [review]
one small (non-blocking) wording change

File: src/declarations/stencil-public-docs.ts (hunk 1 above)
Comment (rwaskiewicz) [reply src/declarations/stencil-public-docs.ts:23]: Bah! How did I miss this? I'll put a tiny followup PR for this. Sorry about that!

**Pull request merged**